# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxTextRun.h
# Commit: 095b3edec3c8
# Full Hash: 095b3edec3c8ba9318c71d0acc91e10ce57818dc
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2019-04-29 21:53:38
# Regressor Bug: 1514869
# File Overlap Count: 1
# Description:
#   Bug 1514869 - patch 2 - Adapt platform-font-list code to work with either the existing in-process font list or cross-process shared font list. r=jwatt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D22938
# ==============================================================================

diff -r 21ef00977ab6 -r 095b3edec3c8 gfx/thebes/gfxTextRun.h
--- a/gfx/thebes/gfxTextRun.h	Sat Apr 27 15:37:29 2019 +0000
+++ b/gfx/thebes/gfxTextRun.h	Sat Apr 27 15:37:58 2019 +0000
@@ -15,6 +15,7 @@
 #include "gfxFontConstants.h"
 #include "gfxSkipChars.h"
 #include "gfxPlatform.h"
+#include "gfxPlatformFontList.h"
 #include "mozilla/MemoryReporting.h"
 #include "mozilla/RefPtr.h"
 #include "nsPoint.h"
@@ -40,6 +41,7 @@
 class gfxMissingFontRecorder;
 
 namespace mozilla {
+class PostTraversalTask;
 class SVGContextPaint;
 enum class StyleHyphens : uint8_t;
 };  // namespace mozilla
@@ -991,6 +993,8 @@
       LazyReferenceDrawTargetGetter& aRefDrawTargetGetter);
 
  protected:
+  friend class mozilla::PostTraversalTask;
+
   struct TextRange {
     TextRange(uint32_t aStart, uint32_t aEnd, gfxFont* aFont,
               FontMatchType aMatchType,
@@ -1022,101 +1026,182 @@
   class FamilyFace {
    public:
     FamilyFace()
-        : mFamily(nullptr),
+        : mOwnedFamily(nullptr),
           mFontEntry(nullptr),
           mGeneric(mozilla::StyleGenericFontFamily::None),
           mFontCreated(false),
           mLoading(false),
           mInvalid(false),
-          mCheckForFallbackFaces(false) {}
+          mCheckForFallbackFaces(false),
+          mIsSharedFamily(false),
+          mHasFontEntry(false) {}
 
     FamilyFace(gfxFontFamily* aFamily, gfxFont* aFont,
                mozilla::StyleGenericFontFamily aGeneric)
-        : mFamily(aFamily),
+        : mOwnedFamily(aFamily),
           mGeneric(aGeneric),
           mFontCreated(true),
           mLoading(false),
           mInvalid(false),
-          mCheckForFallbackFaces(false) {
+          mCheckForFallbackFaces(false),
+          mIsSharedFamily(false),
+          mHasFontEntry(false) {
       NS_ASSERTION(aFont, "font pointer must not be null");
       NS_ASSERTION(!aFamily || aFamily->ContainsFace(aFont->GetFontEntry()),
                    "font is not a member of the given family");
+      NS_IF_ADDREF(aFamily);
       mFont = aFont;
       NS_ADDREF(aFont);
     }
 
     FamilyFace(gfxFontFamily* aFamily, gfxFontEntry* aFontEntry,
                mozilla::StyleGenericFontFamily aGeneric)
-        : mFamily(aFamily),
+        : mOwnedFamily(aFamily),
           mGeneric(aGeneric),
           mFontCreated(false),
           mLoading(false),
           mInvalid(false),
-          mCheckForFallbackFaces(false) {
+          mCheckForFallbackFaces(false),
+          mIsSharedFamily(false),
+          mHasFontEntry(true) {
       NS_ASSERTION(aFontEntry, "font entry pointer must not be null");
       NS_ASSERTION(!aFamily || aFamily->ContainsFace(aFontEntry),
                    "font is not a member of the given family");
+      NS_IF_ADDREF(aFamily);
+      mFontEntry = aFontEntry;
+      NS_ADDREF(aFontEntry);
+    }
+
+    FamilyFace(mozilla::fontlist::Family* aFamily, gfxFontEntry* aFontEntry,
+               mozilla::StyleGenericFontFamily aGeneric)
+        : mSharedFamily(aFamily),
+          mGeneric(aGeneric),
+          mFontCreated(false),
+          mLoading(false),
+          mInvalid(false),
+          mCheckForFallbackFaces(false),
+          mIsSharedFamily(true),
+          mHasFontEntry(true) {
+      MOZ_ASSERT(aFamily && aFontEntry && aFontEntry->mShmemFace);
       mFontEntry = aFontEntry;
       NS_ADDREF(aFontEntry);
     }
 
     FamilyFace(const FamilyFace& aOtherFamilyFace)
-        : mFamily(aOtherFamilyFace.mFamily),
-          mGeneric(aOtherFamilyFace.mGeneric),
+        : mGeneric(aOtherFamilyFace.mGeneric),
           mFontCreated(aOtherFamilyFace.mFontCreated),
           mLoading(aOtherFamilyFace.mLoading),
           mInvalid(aOtherFamilyFace.mInvalid),
-          mCheckForFallbackFaces(aOtherFamilyFace.mCheckForFallbackFaces) {
-      if (mFontCreated) {
-        mFont = aOtherFamilyFace.mFont;
-        NS_ADDREF(mFont);
+          mCheckForFallbackFaces(aOtherFamilyFace.mCheckForFallbackFaces),
+          mIsSharedFamily(aOtherFamilyFace.mIsSharedFamily),
+          mHasFontEntry(aOtherFamilyFace.mHasFontEntry) {
+      if (mIsSharedFamily) {
+        mSharedFamily = aOtherFamilyFace.mSharedFamily;
+        if (mFontCreated) {
+          mFont = aOtherFamilyFace.mFont;
+          NS_ADDREF(mFont);
+        } else if (mHasFontEntry) {
+          mFontEntry = aOtherFamilyFace.mFontEntry;
+          NS_ADDREF(mFontEntry);
+        } else {
+          mSharedFace = aOtherFamilyFace.mSharedFace;
+        }
       } else {
-        mFontEntry = aOtherFamilyFace.mFontEntry;
-        NS_IF_ADDREF(mFontEntry);
+        mOwnedFamily = aOtherFamilyFace.mOwnedFamily;
+        NS_IF_ADDREF(mOwnedFamily);
+        if (mFontCreated) {
+          mFont = aOtherFamilyFace.mFont;
+          NS_ADDREF(mFont);
+        } else {
+          mFontEntry = aOtherFamilyFace.mFontEntry;
+          NS_IF_ADDREF(mFontEntry);
+        }
       }
     }
 
     ~FamilyFace() {
       if (mFontCreated) {
         NS_RELEASE(mFont);
-      } else {
-        NS_IF_RELEASE(mFontEntry);
+      }
+      if (!mIsSharedFamily) {
+        NS_IF_RELEASE(mOwnedFamily);
+      }
+      if (mHasFontEntry) {
+        NS_RELEASE(mFontEntry);
       }
     }
 
     FamilyFace& operator=(const FamilyFace& aOther) {
       if (mFontCreated) {
         NS_RELEASE(mFont);
-      } else {
-        NS_IF_RELEASE(mFontEntry);
+      }
+      if (!mIsSharedFamily) {
+        NS_IF_RELEASE(mOwnedFamily);
+      }
+      if (mHasFontEntry) {
+        NS_RELEASE(mFontEntry);
       }
 
-      mFamily = aOther.mFamily;
       mGeneric = aOther.mGeneric;
       mFontCreated = aOther.mFontCreated;
       mLoading = aOther.mLoading;
       mInvalid = aOther.mInvalid;
+      mIsSharedFamily = aOther.mIsSharedFamily;
+      mHasFontEntry = aOther.mHasFontEntry;
 
-      if (mFontCreated) {
-        mFont = aOther.mFont;
-        NS_ADDREF(mFont);
+      if (mIsSharedFamily) {
+        mSharedFamily = aOther.mSharedFamily;
+        if (mFontCreated) {
+          mFont = aOther.mFont;
+          NS_ADDREF(mFont);
+        } else if (mHasFontEntry) {
+          mFontEntry = aOther.mFontEntry;
+          NS_ADDREF(mFontEntry);
+        } else {
+          mSharedFace = aOther.mSharedFace;
+        }
       } else {
-        mFontEntry = aOther.mFontEntry;
-        NS_IF_ADDREF(mFontEntry);
+        mOwnedFamily = aOther.mOwnedFamily;
+        NS_IF_ADDREF(mOwnedFamily);
+        if (mFontCreated) {
+          mFont = aOther.mFont;
+          NS_ADDREF(mFont);
+        } else {
+          mFontEntry = aOther.mFontEntry;
+          NS_IF_ADDREF(mFontEntry);
+        }
       }
 
       return *this;
     }
 
-    gfxFontFamily* Family() const { return mFamily.get(); }
+    gfxFontFamily* OwnedFamily() const {
+      MOZ_ASSERT(!mIsSharedFamily);
+      return mOwnedFamily;
+    }
+    mozilla::fontlist::Family* SharedFamily() const {
+      MOZ_ASSERT(mIsSharedFamily);
+      return mSharedFamily;
+    }
     gfxFont* Font() const { return mFontCreated ? mFont : nullptr; }
 
     gfxFontEntry* FontEntry() const {
-      return mFontCreated ? mFont->GetFontEntry() : mFontEntry;
+      if (mFontCreated) {
+        return mFont->GetFontEntry();
+      }
+      if (mHasFontEntry) {
+        return mFontEntry;
+      }
+      if (mIsSharedFamily) {
+        return gfxPlatformFontList::PlatformFontList()->GetOrCreateFontEntry(
+            mSharedFace, SharedFamily());
+      }
+      return nullptr;
     }
 
     mozilla::StyleGenericFontFamily Generic() const { return mGeneric; }
 
+    bool IsSharedFamily() const { return mIsSharedFamily; }
     bool IsUserFontContainer() const {
       return FontEntry()->mIsUserFontContainer;
     }
@@ -1133,8 +1218,9 @@
       NS_ADDREF(aFont);
       if (mFontCreated) {
         NS_RELEASE(mFont);
-      } else {
-        NS_IF_RELEASE(mFontEntry);
+      } else if (mHasFontEntry) {
+        NS_RELEASE(mFontEntry);
+        mHasFontEntry = false;
       }
       mFont = aFont;
       mFontCreated = true;
@@ -1144,19 +1230,25 @@
     bool EqualsUserFont(const gfxUserFontEntry* aUserFont) const;
 
    private:
-    RefPtr<gfxFontFamily> mFamily;
+    union {
+      gfxFontFamily* MOZ_OWNING_REF mOwnedFamily;
+      mozilla::fontlist::Family* MOZ_NON_OWNING_REF mSharedFamily;
+    };
     // either a font or a font entry exists
     union {
       // Whichever of these fields is actually present will be a strong
       // reference, with refcounting handled manually.
       gfxFont* MOZ_OWNING_REF mFont;
       gfxFontEntry* MOZ_OWNING_REF mFontEntry;
+      mozilla::fontlist::Face* MOZ_NON_OWNING_REF mSharedFace;
     };
     mozilla::StyleGenericFontFamily mGeneric;
     bool mFontCreated : 1;
     bool mLoading : 1;
     bool mInvalid : 1;
     bool mCheckForFallbackFaces : 1;
+    bool mIsSharedFamily : 1;
+    bool mHasFontEntry : 1;
   };
 
   // List of font families, either named or generic.
@@ -1223,7 +1315,7 @@
 
   // Whether there's a font loading for a given family in the fontlist
   // for a given character
-  bool FontLoadingForFamily(gfxFontFamily* aFamily, uint32_t aCh) const;
+  bool FontLoadingForFamily(const FamilyFace& aFamily, uint32_t aCh) const;
 
   // will always return a font or force a shutdown
   gfxFont* GetDefaultFont();
@@ -1251,6 +1343,11 @@
   // Helper for font-matching:
   // search all faces in a family for a fallback in cases where it's unclear
   // whether the family might have a font for a given character
+  gfxFont* FindFallbackFaceForChar(const FamilyFace& aFamily, uint32_t aCh);
+
+  gfxFont* FindFallbackFaceForChar(mozilla::fontlist::Family* aFamily,
+                                   uint32_t aCh);
+
   gfxFont* FindFallbackFaceForChar(gfxFontFamily* aFamily, uint32_t aCh);
 
   // helper methods for looking up fonts
@@ -1262,6 +1359,8 @@
   // do style selection and add entries to list
   void AddFamilyToFontList(gfxFontFamily* aFamily,
                            mozilla::StyleGenericFontFamily aGeneric);
+  void AddFamilyToFontList(mozilla::fontlist::Family* aFamily,
+                           mozilla::StyleGenericFontFamily aGeneric);
 };
 
 // A "missing font recorder" is to be used during text-run creation to keep